File: Problems/0404_Sum_of_Leaves/Sum_of_Leaves.cs
using System;
using System.Collections.Generic;

// Definition for a binary tree node.
 public class TreeNode {
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int x) { val = x; }
}

public class Solution
{
    public int SumOfLeftLeaves(TreeNode root)
    {
        if (root == null)
            return 0;
        int sum = 0;
        if (root.left != null)
            if (root.left.left == null && root.left.right == null)
                sum += root.left.val;
            else
                sum += sub_sumOfLeftLeaves(root.left);
        if (root.right != null)
            sum += sub_sumOfLeftLeaves(root.right);
        return sum;
    }
    
    public int sub_sumOfLeftLeaves(TreeNode node)
    {
        int sum = 0;
        if (node.left != null)
            if (node.left.left == null && node.left.right == null)
                sum = node.left.val;
            else
                sum += sub_sumOfLeftLeaves(node.left);
        if (node.right != null)
            sum += sub_sumOfLeftLeaves(node.right);
        return sum;
    }

    private TreeNode set_node(string[] flds, int depth, int pos)
    {
        if (flds.Length == 0)
            return null;

        int cur_pos = 0;
        for (int i = 0; i < depth; ++i)
            cur_pos += (int)Math.Pow(2, i);
        
        if (cur_pos + pos > flds.Length - 1)
            return null;
        
        if (flds[cur_pos + pos] == "null")
            return null;

        TreeNode node = new TreeNode(int.Parse(flds[cur_pos + pos]));
        node.left = set_node(flds, depth + 1, 2*pos);
        node.right = set_node(flds, depth + 1, 2*pos + 1);

        return node;
    }

    public string output_int_array(int[] nums)
    {
        if (nums.Length <= 0)
            return "";

        string resultStr = nums[0].ToString();

        for (int i = 1; i < resultStr.Length; ++i)
        {
            resultStr += ", " + nums[i].ToString();
        }

        return resultStr;
    }

    public void Main(string args)
    {
        Console.WriteLine("args = " + args );
        string flds = args.Replace("\"", "").Replace("[", "").Replace("]", "").Trim();
        string[] nums1 = flds.Split(',');

        TreeNode root = set_node(nums1, 0, 0);
        Output_TreeNode ot = new Output_TreeNode();

        Console.WriteLine("root = \n" + ot.output(root));

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        int result = SumOfLeftLeaves(root);
        Console.WriteLine("result = " + result.ToString());

        sw.Stop();
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms");
    }
}

public class Output_TreeNode
{
    List<string> resultStr;
    public string output(TreeNode node)
    {
        resultStr = new List<string>();

        output_tree(node, 0);
        return print_result();
    }
    public void output_tree(TreeNode node, int n)
    {
        if (node == null)
            return;
        
        if (resultStr.Count <= n)
            resultStr.Add("(" + node.val.ToString() + ")");
        else
            resultStr[n] += ",(" + node.val.ToString() + ")";

        if (node.left != null)
            output_tree(node.left, n + 1);
        if (node.right != null)
            output_tree(node.right, n + 1);

        return;
    }

    public string print_result()
    {
        string outputStr = "";

        for (int i = 0; i < resultStr.Count; ++i)
        {
            outputStr += resultStr[i] + "\n";
        }

        resultStr = null;
        return outputStr;
    }
}

File: Problems/0404_Sum_of_Leaves/Project_C/Sum_of_Leaves.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <math.h>
#include <time.h>

#include "mylib.h"

/* Definition for a binary tree node.*/
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

/* Function prototype declaration */
int sumOfLeftLeaves(struct TreeNode* root);
int sub_sumOfLeftLeaves(struct TreeNode* node);

struct TreeNode *set_node(char *flds[], int flds_length, int depth, int pos);
void output_tree(struct TreeNode *node);
void output_node(struct TreeNode *node, char *resultStr[], int n);
void str_replace(const char *src, const char *target, const char *replace, char **result);
int loop_main(char *arg);


int sumOfLeftLeaves(struct TreeNode* root)
{
    if (root == NULL)
        return 0;
    int sum = 0;
    if (root->left != NULL)
        if (root->left->left == NULL && root->left->right == NULL)
            sum += root->left->val;
        else
            sum += sub_sumOfLeftLeaves(root->left);
    if (root->right != NULL)
        sum += sub_sumOfLeftLeaves(root->right);
    return sum;
}
    
int sub_sumOfLeftLeaves(struct TreeNode* node)
{
    int sum = 0;
    if (node->left != NULL)
        if (node->left->left == NULL && node->left->right == NULL)
            sum = node->left->val;
        else
            sum += sub_sumOfLeftLeaves(node->left);
    if (node->right != NULL)
        sum += sub_sumOfLeftLeaves(node->right);
    return sum;
}

struct TreeNode *set_node(char *flds[], int flds_length, int depth, int pos)
{
    if ( flds_length == 0)
        return NULL;

    int cur_pos = 0;
    for (int i = 0; i < depth; ++i)
        cur_pos += (int)pow(2, i);
    
    if (cur_pos + pos > flds_length - 1)
        return NULL;
    
    if (flds[cur_pos + pos] == "null")
        return NULL;

    struct TreeNode *node = malloc(sizeof(struct TreeNode));
    node->val = strtol(flds[cur_pos + pos], NULL, 10);
    node->left = set_node(flds, flds_length, depth + 1, 2*pos);
    node->right = set_node(flds, flds_length, depth + 1, 2*pos + 1);

    return node;
}

void output_tree(struct TreeNode *node)
{
    char *resultStr[256];

    for (int i = 0; i < 256; ++i) {
        resultStr[i] = malloc(sizeof(char)*256);
        resultStr[i][0] = '\0';
    }

    output_node(node, resultStr, 0);

    for (int i = 0; i < 256 && resultStr[i][0] != '\0'; ++i)
        printf("%s\n", resultStr[i]);
}

void output_node(struct TreeNode *node, char *resultStr[], int n)
{
    if (node == NULL)
        return;
    
    char val_to_str[256];
    if (resultStr[n][0] == '\0')
        sprintf(val_to_str, "(%d)", node->val);
    else
        sprintf(val_to_str, ",(%d)", node->val);

    strcat(resultStr[n], val_to_str);
    
    if (node->left != NULL) {
        output_node(node->left, resultStr, n + 1);
    }
    if (node->right != NULL) {
        output_node(node->right, resultStr, n + 1);
    }
}

int loop_main(char *arg)
{
    int argv_length = strlen(arg);
    char tempstr1[256];
    char tempstr2[256];
    char *flds[2];
    char *nums1[256];
    char *nums2[256];

    replace(arg, "[", "");
    replace(arg, "]", "");
    replace(arg, "\n", "");

    int flds_length = split(arg, ",", flds);
    struct TreeNode *root = set_node(flds, flds_length, 0, 0);

    printf("root = \n");
    clock_t time_start = clock();

    output_tree(root);

    clock_t time_end = clock();

    printf("result = %d\n", sumOfLeftLeaves(root));

    printf("Execute time ... %.0f ms\n", 1000*(double)(time_end - time_start)/CLOCKS_PER_SEC);
}

int main(int argc, char *argv[])
{
    FILE *fp;
    char str[256];

    if (argc < 2) {
        printf("Usage %s <testdatafile>\n", argv[0]);
        return -1;
    }

    // File Open
    fp = fopen(argv[1], "r");

    if (fp == NULL) {
        printf("%s Open Failed...\n", argv[1]);
        return -1;
    }

    while((fgets(str, 256, fp)) != NULL) {
        printf("arg = %s\n", str);
        loop_main(str);
    }

    fclose(fp);
    return 0;
}